### DISCUSSION TITLE
Error while executing grab_prefixes

### DISCUSSION THREAD

**Discussion by daverdzx** [peering-manager/peering-manager#676] [Q&A]
Hello Guys hope you are doing well.  Please can you help me, I'm trying to execute the grab_prefixes command as follows, but after a couple of minutes of execution PMGR gets this error:

root@peeringmanager:/opt/peering-manager# source venv/bin/activate
root@peeringmanager:/opt/peering-manager# venv/bin/python3 manage.py grab_prefixes --limit 500
[*] Getting prefixes for AS with IRR AS-SETs
Traceback (most recent call last):
  File "manage.py", line 11, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/base.py", line 414, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/base.py", line 460, in execute
    output = self.handle(*args, **options)
  File "/opt/peering-manager/peering/management/commands/grab_prefixes.py", line 52, in handle
    raise (e)
  File "/opt/peering-manager/peering/management/commands/grab_prefixes.py", line 26, in handle
    prefixes = autonomous_system.retrieve_irr_as_set_prefixes()
  File "/opt/peering-manager/peering/models/models.py", line 302, in retrieve_irr_as_set_prefixes
    call_irr_as_set_resolver(f"AS{self.asn}", address_family=6)
  File "/opt/peering-manager/peering/__init__.py", line 46, in call_irr_as_set_resolver
    raise ValueError(error_log)
ValueError: bgpq3 exit code is 1, stderr: b"ERROR:Invalid source(s) 'RIPE,APNIC,AFRINIC,ARIN,NTTCOM,ALTDB,BBOI,BELL,JPIRR,LEVEL3,RADB,RGNET,TC': \n"
(venv) root@peeringmanager:/opt/peering-manager# 

Note:  I'am executing the manage.py peeringdb_sync periodically and this command works just fine, but when I run the grab_prefixes command well I get the error mentioned above.

Thanks a lot, I appreciate your help.

**Comment by gmazoyer** [comment]
As the error message mentions it, there is an issue with the sources used for looking up prefixes.

You can fix this by changed the `BGPQ3_SOURCES` setting value to something else. It must be a single string with sources separated by `,`.

If you find which source(s) is/are invalid, I can remove them from the default config for next release. Maybe we should limit the default value to authoritative IRR databases `AFRINIC,ARIN,APNIC,LACNIC,RIPE`.

**Reply by daverdzx** [reply]
Hello guys, thanks for your help. I changed the BGPQ3_SOURCES to
BGPQ3_SOURCES = 'AFRINIC,ARIN,APNIC,LACNIC,RIPE'
And tried with different combinations and Im getting the same error:

The following is with BGPQ3_SOURCES = 'ARIN' only:

(venv) root@peeringmanager:/opt/peering-manager# venv/bin/python3 manage.py grab_prefixes --limit 50
[*] Getting prefixes for AS with IRR AS-SETs
Traceback (most recent call last):
File "manage.py", line 11, in
execute_from_command_line(sys.argv)
File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/init.py", line 446, in execute_from_command_line
utility.execute()
File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/init.py", line 440, in execute
self.fetch_command(subcommand).run_from_argv(self.argv)
File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/base.py", line 414, in run_from_argv
self.execute(*args, **cmd_options)
File "/opt/peering-manager/venv/lib/python3.8/site-packages/django/core/management/base.py", line 460, in execute
output = self.handle(*args, **options)
File "/opt/peering-manager/peering/management/commands/grab_prefixes.py", line 52, in handle
raise (e)
File "/opt/peering-manager/peering/management/commands/grab_prefixes.py", line 26, in handle
prefixes = autonomous_system.retrieve_irr_as_set_prefixes()
File "/opt/peering-manager/peering/models/models.py", line 305, in retrieve_irr_as_set_prefixes
call_irr_as_set_resolver(f"AS{self.asn}", address_family=4)
File "/opt/peering-manager/peering/init.py", line 46, in call_irr_as_set_resolver
raise ValueError(error_log)
ValueError: bgpq3 exit code is 1, stderr: b"ERROR:Invalid source(s) 'ARIN': \n"

How can ARIN be an invalid source? I mean
I don't know if maybe it has something to do with some ASN parameter (eg: ASN number or IRR-AS-SET) that is not valid or cannot be resolved via any of the IRR databases.

How can I get more logs of what is happening? I added DEBUG = True in "configuration.py" restarted pmgr process, gave full permissions to "/opt/peering-manager/logs" but no logs are written.

Thank you for your help, I appreciate it.

**Reply by gmazoyer** [reply]
If my memory serves me right, AS are sorted by ASN and the command goes through each AS one by one. So the first AS for which prefixes are being fetched is the one with the lowest ASN. Given your output, it looks like you don't see even the first AS, so it'll mean `bgpq3` fails to retrieve the prefixes for that very first AS.

Sadly you won't be able to get more logs as `bgpq3` exits with an exit code different from 0. You can try to run the `bgpq3` command manually to see if it fails as well.

**Reply by daverdzx** [reply]
Hello Guys thanks for your help, hope you are well.
I was performing some tests, and yes it seems like the AS are fetched in order from lowest to highest by BGPQ3 grab_prefixes, and it seems like it is breaking or getting the error after fetching and getting results successfully for 224 consecutive ASNs.

Here is a fragment of a query output for each asn in order:

... Starts with first asns all with prefixes ok
... then continues as follows:

ASN: 2828
{'ipv4': [{'exact': False, 'prefix': '23.33.32.0/22', 'less-equal': 24}, {'exact': True, 'prefix': '64.3.60.0/24'}, {'exact': True, 'prefix': '64.244.142.0/24'}, {'exact': True, 'prefix': '104.77.153.0/24'}, {'exact': True, 'prefix': '104.78.78.0/24'}, {'exact': True, 'prefix': '209.31.80.0/24'}], 'ipv6': []}

ASN: 2904
{'ipv4': [{'exact': False, 'prefix': '148.210.0.0/16', 'less-equal': 24}], 'ipv6': []}
... 
more ok results
 ... 

ASN: 22011     <<<< seems like its breaking here
None

ASN: 22356
None

ASN: 22381
None

ASN: 22566
None

ASN: 22616
None
.... and until the end only gets prefixes = "None" for each ASN

Executing bgpq3 manually looks like it has no problem with AS22011

(venv) root@peeringmanager:/opt/peering-manager# bgpq3 -Al mypfx AS22011
no ip prefix-list mypfx
ip prefix-list mypfx permit 131.100.0.0/24
ip prefix-list mypfx permit 131.100.3.0/24
ip prefix-list mypfx permit 143.137.108.0/22
...
(venv) root@peeringmanager:/opt/peering-manager# bgpq3 -6 -Al mypfx AS22011
no ipv6 prefix-list mypfx
ipv6 prefix-list mypfx permit 2001:1270::/32
ipv6 prefix-list mypfx permit 2806:23b::/32

And from the GUI AS22011 looks like it is synced correctly to PeeringDB.

Note:  If I generate the prefix-lists via GUI / jinja template with the "prefixes = as | prefix_list" command seems like everything is working just fine because I run this for every router and have no problems getting the prefix-lists configuration, the only drawback is that some large prefix-lists takes more time to be generated.  That is the reason I was trying to optimize time consumption for pfx-lists generation pre-populating the DB with the grab_prefixes command, sadly I am encountering what I just explained in this discussion.  Im gonna keep looking and try to understand more what is happening, any help or hint is really appreciated.

Thank you guys.

**Reply by gmazoyer** [reply]
Trying to get prefixes for AS22011 looks OK to me with the default config.

```
[*] Fetching prefixes for autonomous systems
  - AS22011:
         2 ipv6
        26 ipv4
```

Out of curiosity, what version of Peering Manager are you using?

**Reply by daverdzx** [reply]
Hello Guys, I am using Peering Manager version (v1.7.0)

Regards!
